src/axom/primal/tests/primal_closest_point.cpp

```diff
@@ -11,6 +11,118 @@
+  EXPECT_TRUE(primal::closest_point(QPoint({0.5e-10, 1.0, 0.0}), S, t, EPS) == A);
+  EXPECT_NEAR(t, 0.0, EPS);
+
+  // Query point is equal to A
+  EXPECT_TRUE(primal::closest_point(QPoint({ 0.0, 0.0, 0.0}), S, t, EPS) == A);
+  EXPECT_NEAR(t, 0.0, EPS);
+
+  // Query point is equal to B (for a degenerate segment, A is always returned)
+  EXPECT_TRUE(primal::closest_point(QPoint({ 1.0e-9, 0.0, 0.0}), S, t, EPS) == A);
+  EXPECT_NEAR(t, 0.0, EPS);
+}
+
+//------------------------------------------------------------------------------
+TEST(primal_closest_point, seg_test_closest_point_vertex_0)
+{
+  constexpr double EPS = 1e-8;
+
+  constexpr int DIM = 3;
+  using CoordType = double;
+  using QPoint = primal::Point<CoordType, DIM>;
+  using QSegment = primal::Segment<CoordType, DIM>;
+
+  QPoint A({0.0, 1.0, 1.0});
+  QPoint B({1.0, 0.0, 0.0});
+  QSegment S(A, B);
+  double t;
+
+  // Query point is on the line extended past point A
+  EXPECT_TRUE(primal::closest_point(QPoint({-1.0, 2.0, 2.0}), S, t, EPS) == A);
+  EXPECT_NEAR(t, 0.0, EPS);
+
+  // Query point is on the line perpendicular to AB running through point A
+  EXPECT_TRUE(primal::closest_point(QPoint({-1.0, 0.0, 2.0}), S, t, EPS) == A);
+  EXPECT_NEAR(t, 0.0, EPS);
+
+  // Query point is equal to A
+  EXPECT_TRUE(primal::closest_point(QPoint({ 0.0, 1.0, 1.0}), S, t, EPS) == A);
+  EXPECT_NEAR(t, 0.0, EPS);
+}
+
+//------------------------------------------------------------------------------
+TEST(primal_closest_point, seg_test_closest_point_vertex_1)
+{
+  constexpr double EPS = 1e-8;
+
+  constexpr int DIM = 3;
+  using CoordType = double;
+  using QPoint = primal::Point<CoordType, DIM>;
+  using QSegment = primal::Segment<CoordType, DIM>;
+
+  QPoint A({0.0, 1.0, 1.0});
+  QPoint B({1.0, 0.0, 0.0});
+  QSegment S(A, B);
+  double t;
+
+  // Query point is on the line extended past point B
+  EXPECT_TRUE(primal::closest_point(QPoint({ 2.0, -1.0,  1.0}), S, t, EPS) == B);
+  EXPECT_NEAR(t, 1.0, EPS);
+
+  // Query point is on the line perpendicular to AB running through point B
+  EXPECT_TRUE(primal::closest_point(QPoint({ 2.0,  1.0, -1.0}), S, t, EPS) == B);
+  EXPECT_NEAR(t, 1.0, EPS);
+
+  // Query point is equal to B
+  EXPECT_TRUE(primal::closest_point(QPoint({ 1.0,  0.0,  0.0}), S, t, EPS) == B);
+  EXPECT_NEAR(t, 1.0, EPS);
+}
+
+//------------------------------------------------------------------------------
+TEST(primal_closest_point, seg_test_closest_point_interior)
+{
+  constexpr double EPS = 1e-8;
+
+  constexpr int DIM = 3;
+  using CoordType = double;
+  using QPoint = primal::Point<CoordType, DIM>;
+  using QSegment = primal::Segment<CoordType, DIM>;
+
+  QPoint A({0.0, 1.0, 1.0});
+  QPoint B({1.0, 0.0, 0.0});
+  QSegment S(A, B);
+  double t;
+
+  // Query point is perpendicular to the midpoint of AB
+  EXPECT_TRUE(primal::closest_point(QPoint({0.0, 0.0, 0.5}), S, t, EPS) == QPoint::lerp(A, B, 0.5));
```

Comment:
Can someone sanity check this particular test case?

Reply:
Seems reasonable to me
$A := (0,1,1)$, $B := (1,0,0)$, $Q := (0,0,.5)$
$AB  := <1, -1, -1>$
$AQ :=  <0,-1,-.5>$
$AB \cdot AB = 3$
$AB \cdot AQ = 1.5$

$t = (AB \cdot AQ)/(AB \cdot AB) = .5$

Reply:
I suppose you can also try comparing the value against a segment with the reverse orientation, which should give you the same answer:

```
  QSegment S_reverse(B, A);
  EXPECT_TRUE(primal::closest_point(QPoint({0.0, 0.0, 0.5}), S_reverse, t, EPS) == QPoint::lerp(A, B, 0.5));
```

Edit: It might not be a bad idea to test the reversed segment in all cases.

Reply:
I added tests for all the reversed segments.